### ISSUE TITLE
tovid todisc line 454   printf 1.2345 invalid number

### ISSUE THREAD

**Issue by grepper** [grepper/tovid#86]
_From [claude.o...@gmail.com](https://code.google.com/u/114500179454331997356/) on December 21, 2008 10:45:03_

bash uses only integers !!!

tovid_demo todisc svn 2777 line 454   printf 1.2345 invalid number
bash line 0 0,0000 invalid number

the result is todisc ending with

A seek of  2 sec is too large for foo.mpg
the file  is only 0,000 secs long
what printf format to use to get integers only ?
printf &#37;d ?

**Attachment:** [tovid-demo.log todisc.log](http://code.google.com/p/tovid/issues/detail?id=85)

_Original issue: http://code.google.com/p/tovid/issues/detail?id=85_

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on December 21, 2008 13:04:32_

This is a localization problem.  ( comma rather than period decimal separator. )
I'll have to look at it.
Thanks for the report.

grepper

**Owner:** grepper  
  **Labels:** Affects-todisc

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on April 25, 2010 20:50:30_

the tovid-demo scripts have been removed for now, as they are difficult to maintain
with the changeability of the internet, from which some files were downloaded.

**Status:** WontFix

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on September 14, 2010 17:26:16_

Localization is still a problem for todisc and perhaps some other scripts.  I'm reopening this.

**Status:** Accepted

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on September 15, 2010 18:35:48_

changed title as this is not really about tovid-demo and is still current for svn

**Summary:** tovid todisc line 454   printf 1.2345 invalid number

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on September 15, 2010 18:38:38_

A user reported that adding:
export LANG='C' before running todisc 'tovid disc'
fixed this for him.
Considering adding:
os.environ['LANG'] = 'C'
to src/tovid

**Comment by grepper** [comment]
_From [sven.her...@gmail.com](https://code.google.com/u/107529408755939848526/) on February 22, 2011 07:06:12_

LANG=C tovid disc

is already enough (in bash at least), but it prevents one from using something like german umlauts and such as titles. 

Therefore os.environ['LANG'] = 'C' does not really fix this completely.

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on February 22, 2011 10:08:28_

Did you try using LC_NUMERIC=C rather than LANG ?

grepper

**Comment by grepper** [comment]
_From [sven.her...@gmail.com](https://code.google.com/u/107529408755939848526/) on February 23, 2011 07:02:55_

I believe I did and it didn't work either.

I do not know too much about encoding videos, but I know some python and I am willing to take a look at it. Where is the file we're after here? If you could give me a starting point?

Thanks,
Sven

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on February 23, 2011 20:20:36_

It would be the bash script src/todisc.  I think if locale fixes will not work it would be best to sanitize the output for the video length with something like ${VAR/,/.}.  And then of course to hunt down other numeric output besides video length that may be affected.  I'll have a look too when I get a chance.

grepper

**Comment by grepper** [comment]
_From [sven.her...@gmail.com](https://code.google.com/u/107529408755939848526/) on February 24, 2011 00:31:36_

I believe it has to do with the function bc_math in line 473 and printf. Your script uses the bash builtin printf, which behaves a bit different than the binary printf:

$ printf "%.03f\n" .333
bash: printf: .333: Ungültige Zahl.
0,000

$ printf "%.03f\n" ,333
0,333

$ LANG=C printf "%.03f\n" ,333
0,333

$ LANG=C printf "%.03f\n" .333
bash: printf: .333: Ungültige Zahl.
0,000

$ /usr/bin/printf "%.03f\n" 0.333
0,333

$ /usr/bin/printf "%.03f\n" 0,333
/usr/bin/printf: 0,333: Wert nicht vollständig konvertiert
0,000

$ LANG=C /usr/bin/printf "%.03f\n" 0.333
0.333

$ LANG=C /usr/bin/printf "%.03f\n" 0,333
/usr/bin/printf: 0,333: value not completely converted
0.000

The main point is that the binary does a printf of .333 (though to 0,333) whereas the builtin fails to do so and spits out the error.

Thus, my suggestion regarding the function would be


\# generic function for using bc for floating point math
\# usage: bc_math "expression" [int]
bc_math()
{
    bc_out=$( bc \<<< "scale=3;$1" 2>/dev/null)
    \# just to be sure
    bc_out=${bc_out//,/.}
    if [[ -n $2 && $2 = "int" ]]; then
        echo ${bc_out%.*}
    else
        \# use the builtin and force period 
        LANG=C /usr/bin/printf  "%.03f" $bc_out
    fi
}

This seems to have fixed it for me whereas with the printf-builtin, the error continues to bug us.

Cheers,
Sven

**Comment by grepper** [comment]
_From [sven.her...@gmail.com](https://code.google.com/u/107529408755939848526/) on February 24, 2011 01:24:22_

correction:

for me it only works when I remove "bc_out=${bc_out//,/.}"

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on July 11, 2011 14:23:09_

I have stayed away from the bug as I don't feel competent to address locale issues, having only ever dealt with en_* or C.  It is also unclear to me: your programs are localized, yes ?  Does bc really output a period separated number ?  Does your localized mplayer,ffmpeg etc output AND expect as input comma delimited numbers ?

What would happen for you if I removed the LANG=C (as well as bc_out=${bc_out//,/.} ) altogether?

grepper

**Comment by grepper** [comment]
_From [captblas...@gmail.com](https://code.google.com/u/110042941665766582899/) on March 11, 2013 05:56:05_

I had the "decimal comma" problem. I find this site too late, but I did some changes by myself based on facts provided in the Mandriva forum.
The "complete" thread is here: http://forum.mandriva.com/en/viewtopic.php?f=86&t=138171

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on March 11, 2013 09:43:36_

Thanks for your report.  What happens if you simply use LC_NUMERIC=C ?

try this for the bc_math function:

bc_math()                                                                       
{                                                                               
    LC_NUMERIC=C                                                                
    bc_out=$(bc \<<< "scale=3;$1" 2>/dev/null)                                   
    if [[ -n $2 && $2 = "int" ]]; then                                          
        echo ${bc_out%.*}                                                       
    else                                                                        
        printf  "%.03f" $bc_out                                                 
    fi                                                                          
}

In other words add the line LC_NUMERIC=C to the top of the function.

thanks,
grepper

**Comment by grepper** [comment]
_From [captblas...@gmail.com](https://code.google.com/u/110042941665766582899/) on March 11, 2013 20:28:03_

In the Mandriva Forum it was reported that "LC_NUMERIC=C" works, but I don't test it yet. I cannot do it now, but I will in a day or two.
I have a question: if you change LC_NUMERIC at the beginning of bc_math (or at the beginning of todisc), how do you restore the value it had before running todisc script?
In my system LC_NUMERIC contains "es_AR.UTF-8" (without quotation marks). If I add "LC_NUMERIC=C", the es_AR value is lost. So, how do I recover it? Ok, I know that I can run "LC_NUMERIC=es_AR.UTF-8" after finishing with tovid. But I'm trying to have a solution that it's more "universal", in the sense that it can be used if your LC_NUMERIC is en_GB, mine is es_AR or other one have fr_FR or so. My idea is to make tovid/todisc independent of the LC_NUMERIC value, and LC_NUMERIC (or any other "system variable") not altered because of the execution of tovid/todisc.
Please take into account that I'm not a bash expert at all. Maybe your solution change the value of LC_NUMERIC only "locally" (i.e., its value change only while the todisc script is running, but it recover its original value when the script ends). I cannot test "LC_NUMERIC=C" yet, but when I will, I'll find out if LC_NUMERIC is altered or not (I'm pretty sure that tovid will run OK and the DVD it will be created after modifying bc_math with your or mine solution).

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on March 11, 2013 22:06:20_

Hi,
Thanks for agreeing to test.  It is difficult for me to do so as I have only English language locales installed, and I have little experience with other locales. Though of course I did generate some other locales and did a test of the above function.

As far as your question is concerned: a child process (in this case todisc) can not affect the environment of its parent, so your LC_NUMERIC will be unchanged when the script exits. In addition, I planned to make the assignment local to the function ("local LC_NUMERIC"), so it will not even affect other parts of the script which will still run with your systems LC_NUMERIC. If that is not what you want, it can be exported at the top of the script to affect it globally.

To be honest I am surprised that todisc works with other locales - how does the output with a period from printf now play with programs that expect a comma, like for example time codes given to mplayer or ffmpeg ?
As you can see, I am pretty foggy about these issues.

**Comment by grepper** [comment]
_From [captblas...@gmail.com](https://code.google.com/u/110042941665766582899/) on March 13, 2013 21:10:43_

I just test the simpler solution.

LC_NUMERIC=C works, DVD is created, and it's a good enough solution as LC_NUMERIC keeps its value after todisc ends execution. I put the "solution" in the same thread of Mandriva forum ( http://forum.mandriva.com/en/viewtopic.php?f=86&t=138171 )

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on March 15, 2013 21:37:36_

Great, thanks for testing.  There is a subtle bug that could come up here though. If a user has LC_ALL set, then it is not possible to set LC_NUMERIC: the setting for LC_ALL will override it. So here is the new function, which saves LC_ALL, then sets it to "C", then sets it back to its saved value.

bc_math()
{
    OLC_ALL=$LC_ALL
    LC_ALL=C
    bc_out=$(bc \<<< "scale=3;$1" 2>/dev/null)
    if [[ -n $2 && $2 = "int" ]]; then
        echo ${bc_out%.*}
    else
        printf  "%.03f" $bc_out
    fi
    LC_ALL=$OLC_ALL
}

It is a better solution and how it should have been to begin with I think.  After the function runs, LC_ALL will be the same as originally (unset if unset to begin with as it should be) and LC_NUMERIC will be untouched.

Could you please test this?  

cheers,
grepper

**Comment by grepper** [comment]
_From [captblas...@gmail.com](https://code.google.com/u/110042941665766582899/) on March 17, 2013 15:36:45_

It works. LC_NUMERIC and LC_ALL kept their values (well, LC_ALL wasn't defined in my system)
So that's how that bc_math() must be.

**Comment by grepper** [comment]
_From [grepper@gmail.com](https://code.google.com/u/grepper@gmail.com/) on March 17, 2013 19:02:04_

Thanks very much. This issue could have been closed a long time ago if previous submitters had taken my questions more seriously, as you did.  I will close it after I commit it to HEAD in a few days.
all the best,
grepper

**Status:** Fixed
